libgalois/src/analytics/bfs/bfs.cpp

```diff
@@ -390,21 +391,97 @@ SynchronousDirectOpt(
   delete next_frontier;
 }
 
+template <typename NDType, typename ValueTy>
+void
+InitializeNodeData(const ValueTy value, katana::LargeArray<NDType>* node_data) {
+  katana::do_all(katana::iterate(0ul, node_data->size()), [&](auto n) {
+    (*node_data)[n] = value;
+  });
+}
+
+template <typename NDType, typename ValueTy, typename... Args>
+void
+InitializeNodeData(
+    const ValueTy value, katana::LargeArray<NDType>* node_data, Args... args) {
+  InitializeNodeData(value, node_data);
+  InitializeNodeData(value, args...);
+}
+
+template <typename NDType>
+void
+InitializeGraphNodeData(
+    Graph* graph, const katana::LargeArray<NDType>& node_data) {
+  katana::do_all(katana::iterate(*graph), [&](auto& node) {
+    graph->GetData<BfsNodeParent>(node) = node_data[node];
+  });
+}
+
+void
+ComputeParentByDistance(
+    katana::PropertyGraph& transpose_graph,
+    katana::LargeArray<GNode>* node_parent,
+    const katana::LargeArray<Dist>& node_dist, const GNode source) {
+  (*node_parent)[source] = source;
+  katana::do_all(
+      katana::iterate(transpose_graph),
+      [&](const GNode v) {
+        GNode& v_parent = (*node_parent)[v];
+        for (auto e : transpose_graph.edges(v)) {
```

Comment:
Skip this if distance is infinity. Also skip this if distance is 1 - then the parent is always source.